### PR TITLE
[x] adding json/toml output to generate-summaries

### PR DESCRIPTION
**Pull request by mimoo**

### CODE REVIEW

File: devtools/x/src/generate_summaries.rs [hunk 1]
```diff
@@ -18,6 +18,9 @@ pub struct Args {
     #[structopt(name = "OUT_DIR")]
     /// Directory to output summaries to (default: target/summaries)
     out_dir: Option<PathBuf>,
+    #[structopt(name = "OUTPUT_FORMAT")]
+    /// Output in text, toml or json
+    output_format: Option<String>,
```
Comment (sunshowers) [inline devtools/x/src/generate_summaries.rs:23]: This should be an enum using `arg_enum!`, here's an example: https://github.com/diem/diem/blob/39e4d7a064ea4b9d071c56d9df5099b655c4b8f8/devtools/x/src/cargo/build_args.rs#L7

In general, input validation should occur as close to the edges of the program as possible.

File: devtools/x/src/generate_summaries.rs [hunk 2]
```diff
@@ -77,25 +98,39 @@ fn write_summary(
...
+    let (out, path) = match output_format {
+        Some("json") => {
+            let out = serde_json::to_string(&summary)?;
+            let summary_path = out_dir.join(format!("summary-{}.json", name));
+            (out, summary_path)
+        }
+        Some("toml") => {
+            let out = toml::to_string(&summary)?;
+            let summary_path = out_dir.join(format!("summary-{}.toml", name));
+            (out, summary_path)
```
Comment (sunshowers) [inline devtools/x/src/generate_summaries.rs:115]: Please continue to use the `write_to_string` method.

**Review by sunshowers — CHANGES_REQUESTED** [review]

File: devtools/x/src/generate_summaries.rs (hunk 2 above)
Comment (mimoo) [reply devtools/x/src/generate_summaries.rs:115]: you mean to have the text at the beginning of the doc? The problem with that is that it makes the json file invalid, but it works with toml so I can have it for toml if that's what you're asking?

Comment (mimoo) [reply devtools/x/src/generate_summaries.rs:115]: oh wait, the default text was already toml 🤦🏻‍♂️

File: devtools/x/src/generate_summaries.rs (hunk 1 above)
Comment (mimoo) [reply devtools/x/src/generate_summaries.rs:23]: done!

File: devtools/x/src/generate_summaries.rs (hunk 2 above)
Comment (mimoo) [reply devtools/x/src/generate_summaries.rs]: done!

File: devtools/x/src/generate_summaries.rs [hunk 3]
```diff
@@ -11,13 +11,24 @@ use std::{
...
+}
 
 #[derive(Debug, StructOpt)]
 pub struct Args {
     #[structopt(name = "OUT_DIR")]
     /// Directory to output summaries to (default: target/summaries)
     out_dir: Option<PathBuf>,
+    #[structopt(name = "OUTPUT_FORMAT")]
+    /// Output in text, toml or json
+    output_format: Option<OutputFormat>,
```
Comment (sunshowers) [inline devtools/x/src/generate_summaries.rs:31]: I think the only supported output formats are toml and json? If so I think you can just use a non-Option enum here and set a default: https://github.com/diem/diem/blob/39e4d7a064ea4b9d071c56d9df5099b655c4b8f8/devtools/x/src/cargo/build_args.rs#L85-L87

Comment (mimoo) [reply devtools/x/src/generate_summaries.rs]: simplified : o tested, it works!

File: devtools/x/src/diff_summary.rs [hunk 4]
```diff
@@ -15,9 +24,9 @@ pub struct Args {
     #[structopt(name = "COMPARE_SUMMARY")]
     /// Path to the comparison summary
     compare_summary: PathBuf,
-    #[structopt(name = "OUTPUT_FORMAT")]
+    #[structopt(name = "OUTPUT_FORMAT", default_value = "Text")]
```
Comment (sunshowers) [inline devtools/x/src/diff_summary.rs:27]: This should also have a `possible_values` argument: https://github.com/diem/diem/blob/39e4d7a064ea4b9d071c56d9df5099b655c4b8f8/devtools/x/src/cargo/build_args.rs#L85

File: devtools/x/src/generate_summaries.rs [hunk 5]
```diff
@@ -11,13 +11,24 @@ use std::{
...
+        Json,
+    }
+}
 
 #[derive(Debug, StructOpt)]
 pub struct Args {
     #[structopt(name = "OUT_DIR")]
     /// Directory to output summaries to (default: target/summaries)
     out_dir: Option<PathBuf>,
+    #[structopt(name = "OUTPUT_FORMAT", default_value = "Toml")]
```
Comment (sunshowers) [inline devtools/x/src/generate_summaries.rs:29]: as should this :)

**Review by sunshowers — APPROVED** [review]
Looks great, thanks! Just needs a couple more changes before it can land.

File: devtools/x/src/diff_summary.rs (hunk 4 above)
Comment (mimoo) [reply devtools/x/src/diff_summary.rs:27]: I don't think we need this for an enum:

```sh
> cargo x generate-summaries out_dir bad_value
error: Invalid value for '<OUTPUT_FORMAT>': valid values: Toml, Json

> cargo x generate-summaries -h
Generate build summaries for important subsets

USAGE:
    x generate-summaries [ARGS]

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

ARGS:
    <OUT_DIR>          Directory to output summaries to (default: target/summaries)
    <OUTPUT_FORMAT>    Output in text, toml or json [default: Toml]
```

Comment (sunshowers) [reply devtools/x/src/diff_summary.rs:26]: Ahh perfect! Thanks.

**Pull request merged**